### PR TITLE
Make UndoRedo use Callables

### LINKED ISSUE #5038: Make `UndoRedo.add_do/undo_method()` use Callable
### Describe the project you are working on

An SVG editing tool

### Describe the problem or limitation you are having in your project

`UndoRedo.add_do/undo_method()` uses an object reference and method name instead of a Callable.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Replace the first two arguments with a Callable argument.

#### Before:
```gdscript
add_do_method(object: Object, method: StringName, ...)
```

#### After:
```gdscript
add_do_method(method: Callable, ...)
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes, you can use the functions' existing syntax.

### Is there a reason why this should be core and not an add-on in the asset library?

This makes two functions that are already in core more convenient to use.

### LINKED ISSUE #26483: UndoRedo's do/undo methods can't go further than 5 arguments
**Godot version:**
b84b015

**Issue description:**
When using UndoRedo's do/undo methods, the number of arguments can't go further than 5, as anything else will be ignored.

**Minimal reproduction project:**
The following code will give an error stating that the `test()` method was only called with 5 arguments.
```
var _undo_redo := UndoRedo.new()

func _ready() -> void:
	_undo_redo.create_action("test")
	_undo_redo.add_do_method(self, "test",
			true, true, true, true, true, true, true)
	_undo_redo.commit_action()

func test(arg1: bool, arg2: bool, arg3: bool, arg4: bool, arg5: bool,
		arg6: bool, arg7: bool) -> void:
	print(arg1, arg2, arg3, arg4, arg5, arg6, arg7)
```

### LINKED ISSUE #36895: It is not possible to send `null` as the last argument of an UndoRedo method
Godot 3.2

To register a method to be executed when doing or undoing a change, `UndoRedo.add_do_method()` or `UndoRedo.add_undo_method()` have to be used.
Those methods were implemented by declaring 5 `Variant` arguments, all `null` by default.
Unfortunately, the argument count is then determined by checking how many remaining arguments are null, *including cases where you want to call the method with `null`*.
That means it is impossible to write this:
```gdscript
undo_redo.add_do_method(object, "foo", 42, true, null)
```
Because the `null` at the end will be confused as absence of argument. Then when committing the action, undoing or redoing it, it won't work and this kind of error will be printed:
```
ERROR: UndoRedo::_process_operation_list: Error calling method from signal 'foo': Method expected 3 arguments, but called with 2.
```
See https://github.com/godotengine/godot/blob/78adce833bce49ef1bf1503956a6162a1c6dd8b2/core/undo_redo.cpp#L284

I had to workaround this limitation by binding extra methods in my objects, and check if variants are null before using `add_do_method` and `add_undo_method`.

To fix this, `UndoRedo` should not determine argument count in this way. Instead, it should use reflection to query how many arguments the called method actually has. Other solutions include sending an array explicitely, adding a count argument, or having a `Variant::UNDEFINED` type, but that would break compatibility.

### PR DESCRIPTION
**Pull request by KoBeWi**
This PR makes UndoRedo use Callables instead of method names and hardcoded argument list. I added `add_undo_method_compat` and changed all old calls to this. We need to port them later and cleanup needlessly bound methods.

Likely fixes #26483
And likely fixes #36895

Sumbitting as draft, because I'm not yet sure what to do with
https://github.com/godotengine/godot/blob/7e009a167110e79caa102c5b88363fd32b75293c/core/object/undo_redo.cpp#L301-L303
Either I need to figure out how to change it to Callable or extract arguments from callable and push them here. For now this part is commented out, everything else seems to work.

*Bugsquad edit: This closes https://github.com/godotengine/godot-proposals/issues/5038.*

### CODE REVIEW

**Comment by Riteo** [comment]
Why does CI fail all jobs except Windows builds and static checks?

**Comment by bruvzg** [comment]
> Why does CI fail all jobs except Windows builds and static checks?

Click "Details" next to the failed job to check build logs.

```
core/object/undo_redo.cpp:308:24: error: non-const lvalue reference to type 'Variant' cannot bind to a temporary of type 'Variant'
                                c.call(nullptr, 0, Variant(), ce);
                                                   ^~~~~~~~~
./core/variant/callable.h:70:66: note: passing argument to parameter 'r_return_value' here
        void call(const Variant **p_arguments, int p_argcount, Variant &r_return_value, CallError &r_call_error) const;
                                                                        ^
1 error generated.
```

**Comment by KoBeWi** [comment]
Ok, finished this. I had to add CallableCustomBind.get_binds() and with some dynamic casting I managed to resolve the callback issue. It was something with live edit, seems to work now. At least when you change properties in the inspector, but when you move nodes it doesn't seem to synchronize properly, no idea why.

**Comment by reduz** [comment]
For the record, I suggest closing this. Using callables in UndoRedo means that contributors will be able to provide function pointers instead of instance/string. This breaks live editing, hence it should not be used.

**Comment by KoBeWi** [comment]
> This breaks live editing, hence it should not be used.

Live editing uses method pointer, so I just extract it from the callable (including arguments, I added a method for that). So live editing still works (except one small issue, see my previous comment).

**Comment by KoBeWi** [comment]
Closing, as there is too many conflicts. Could be revisited after #59564 is merged.

The status is this:
Godot has live scene edit, i.e. when you run the game from the editor, any change you do in the editor (not on remote tree) is synced via debugger to the running scene. This is done via UndoRedo actions - the method name is sent to the running instance and is executed remotely. Method pointer Callables don't provide method names, so they can't be called remotely.
HOWEVER, as I mentioned sometime before, most of the editor changes sent via live edit comes from property changes, which don't use methods. They just send property name and value. Also most editor changes done via methods aren't synced anyway (node renames, animation edits etc). So IMO we could change everything to Callables, but still keep the bindings for methods that need to be synchronized (live edit with named Callables should be possible, because you can get the method name). It's probably a few % of the methods, so it shouldn't be a problem. We'd change all methods to `callable_mp` and use `Callable(object, "method")` for methods that matter for live edit. Although reduz was opposed to doing it last time it was mentioned (you need to know whether the method is relevant for live edit or not and it will confuse users etc)
Another option is that we could keep all bindings and use named callables everywhere. This is rather meh IMO.

tl;dr this can be revived in the future, but should be done carefully to avoid breaking live edit™

**Comment by akien-mga** [comment]
Superseded by #66070.

**Comment by KoBeWi** [comment]
Not really. The current equivalent of this PR would be making EditorUndoRedoManager use callables, but it has the problems I mentioned in the above comment.
